### PR TITLE
Merge polished sharing functionality implementation

### PR DESCRIPTION
**Pull request by sampsapenna**
### Description

Polished sharing functionality implementation to both improve the user experience, and increase security.

### Related issues
* Solves #47 – incorrect breadcrumb behaviour 
* Solves #46 – remove unnecessary detail row in container view
* Solves #45 – make sharing views more accessible with routed view instead of tabs (still not completely accessible though, will be improved in the future)
* Solves #44 – fix shared container access view breaking
* Solves #43 – fix switching views leading to undefined variables due to incorrect use of routing

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
    * Fixes multiple bugs in sharing view implementation and backend
- [x] New feature (non-breaking change which adds functionality)
    * Add API call signing to backend, to enable authenticated CORS requests to external APIs
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
    * Added authentication for sharing and request APIs, leading to the old insecure API being deprecated in the future.

### Changes Made

* Change sharing functionality from tabbed interface -> routed
* Improve sharing functionality styling
* Remove container listing details as unnecessary
* Make Dockerfile build a non-root container
* Improve breadcrumb behavior in the UI
* Parameterize sharing UI text
* Update multiple files in the backend to use type hints (futureproofing)
* Fix undefined client error on page refresh crashing external API access
* Fix tables not updating after altering access requests
* Make sharing functionality rendering dependent on the functionality being enabled in the backend
* Implement external API request signing to the backend using HMAC as signature engine

### Testing

- [x] Needs testing (start an issue or do a follow up PR about it)

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -6,4 +6,4 @@ click
 sphinx
 sphinx_rtd_theme
 uvloop
-gunicorn
\ No newline at end of file
+gunicorn==19.9.0
```
Comment (blankdots) [inline requirements.txt]: I assume the update did not work as well, as the issue should have been fixed.

File: swift_browser_ui_frontend/package.json [hunk 2]
```diff
@@ -3,7 +3,7 @@
   "version": "0.4.0",
```
Comment (blankdots) [inline swift_browser_ui_frontend/package.json]: reminder: version bump would be nice when merging in `master`

File: requirements.txt (hunk 1 above)
Comment (sampsapenna) [reply requirements.txt]: Nope, that was just forgotten from the merge – fixed.

File: swift_browser_ui_frontend/package.json (hunk 2 above)
Comment (sampsapenna) [reply swift_browser_ui_frontend/package.json]: Yes, that's going to happen when the changes are merged from devel->master :+1:

**Pull request merged**